### ISSUE TITLE
Feature Request: Error-prone `case` Pattern

### ISSUE THREAD

**Issue by massimo-zaniboni** [ndmitchell/hlint#295]
This code

```haskell
isLen2 l
  = let is2 = 2
    in case length l of
         is2 -> True
         _ -> False

t = isLen2 []
```
is error-prone because someone, for example me :-) can badly assume this semantic 

```haskell
isLen2' l
  = case length l of
         2 -> True
         _ -> False

t' = isLen2' []
```

but instead the `case` in `f` returns always `True` because (correctly), a new `is2` identifier is defined by the case condition.

I don't know what is the policy for feature-requests, but obviously I'm not pretending nothing, and many many thanks for the very good work on this utility!!!

**Comment by ndmitchell** [comment]
Generally HLint tries to do the things that GHC doesn't do, and GHC does a lot of pattern match checking in its warnings, so it's starting to stray to the boundary where I'd suggest it go in as a GHC warning. That said, perhaps having a sweet spot of warnings to eliminate successive branches if a previous branch matches everything is a good idea.

**Comment by massimo-zaniboni** [comment]
> GHC does a lot of pattern match checking in its warnings,

you are right: for this code GHC already signal a warning. 

By the way I didn't noticed the error in my production code, because it generates some warnings that are instead "false warnings" and so I ignored also this true warning. But this is completely my fault: I had to remove all warning from my code, otherwise these type of things can happen.

IMHO you can close this issue because GHC warned me.

**Comment by ndmitchell** [comment]
I'm not a great fan of GHC's pattern match warnings - they tend to be too dumb to be helpful most of the time - but given they exist and do warn about this I'm going to leave it to GHC HQ to improve them, rather than try and reinvent them in HLint.
